velox/connectors/hive/storage_adapters/s3fs/S3MetricsAggregator.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/storage_adapters/s3fs/S3MetricsAggregator.h"
+#include <folly/Singleton.h>
+#include "velox/common/base/StatsReporter.h"
+#include "velox/connectors/hive/storage_adapters/s3fs/S3Metrics.h"
+namespace facebook::velox::filesystems {
+
+namespace {
+folly::Singleton<S3MetricsAggregator> s3MetricsAggregatorSingleton;
+} // namespace
+
+std::shared_ptr<S3MetricsAggregator> S3MetricsAggregator::getInstance() {
+  return s3MetricsAggregatorSingleton.try_get();
+}
+
+S3MetricsAggregator::S3MetricsAggregator() {
+  // Define and initialize the metrics
+  DEFINE_METRIC(kMetricS3ActiveConnections, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3StartedUploads, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3FailedUploads, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3SuccessfulUploads, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3MetadataCalls, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3ListStatusCalls, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3ListLocatedStatusCalls, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3ListObjectsCalls, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3OtherReadErrors, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3AwsAbortedExceptions, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3SocketExceptions, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3GetObjectErrors, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3GetMetadataErrors, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3GetObjectRetries, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3GetMetadataRetries, velox::StatType::COUNT);
+  DEFINE_METRIC(kMetricS3ReadRetries, velox::StatType::COUNT);
```

Comment:
The DEFINE_METRIC calls need not be in constructor. Move it to velox/common/base/Counters.cpp and put under #ifdef VELOX_ENABLE_S3.

Also, add comments for each of these metrics. Some of the metrics on Java side track the one minute count. In which case they must be of SUM type and not counts.

Worker S3 Started Uploads  |  presto_hive_s3_presto_s3_file_system_started_uploads_one_minute_count
-- | --
Worker S3 Failed Uploads  |  presto_hive_s3_presto_s3_file_system_failed_uploads_one_minute_count
Worker S3 Successful Uploads | presto_hive_s3_presto_s3_file_system_successful_uploads_one_minute_count


While RECORD_METRIC for SUM types make sure we record the delta from the previous recording. Refer other metric instances like cache, spill etc.